bpf,map: Use device map instead of device macro

haozhangami opened the pull request:
This patch reimplements IS_L3_DEV and NATIVE_DEV_MAC_BY_IFINDEX using bpf map, including the following changes:

1.Define a cilium_device_map, this new map will map device ifindexes to their MACs and check if they are L3 devices.
2.Orchestrator manages device entries in the above map.
3.Use is_l3_device() and device_mac() methods to replace IS_L3_DEV and
  NATIVE_DEV_MAC_BY_IFINDEX

Fixes: #39116

```release-note
Reimplement device macros using a new device bpf map
```

julianwiedmann commented on bpf/lib/node.h:
```diff
@@ -65,3 +65,45 @@ lookup_ip6_node_id(const union v6addr *ip6)
+}
+
+static __always_inline union macaddr
+device_mac(__u32 ifindex)
+{
+	union macaddr dev_mac = {0};
+	struct device_value *device_value;
+
+	device_value = lookup_device(ifindex);
+	if (!device_value)
+		return dev_mac;
+	return device_value->mac;
+}
```
`node.h` wraps the node map, which contains *remote* nodes. But now we want to manage network interfaces on the _local_ node. So I think this would be best placed into a separate header file - this could be named `network_devices.h` ?

haozhangami replied:
I have moved this code to a new header file `network_device.h`. Please take a look.

julianwiedmann commented on pkg/maps/devicemap/devicemap.go:
```diff
@@ -0,0 +1,127 @@
+
+const (
+	MapName    = "cilium_device_map"
+	MaxEntries = 16384
```
That's an awfully big number of entries, considering how many network interfaces we'd typically expect to manage :)

haozhangami replied:
Done, reducing the max entries to 256.

julianwiedmann commented on bpf/lib/common.h:
```diff
@@ -447,6 +447,16 @@ struct node_value {
+
+struct device_value {
+	union macaddr   mac;
+	__u8            l3;
+	__u8		pad[7];
```
Did you mean:

```suggestion
	__u8		l3:1;
	__u8		pad:7;
```

haozhangami replied:
pad[7] is for C and GO structs alignment. Without it, alignment check will fail.
```
C and Go structs alignment check in datapath failed: type DeviceValue's binary.Size (9) does not equal its unsafe.Sizeof (16) size
```

julianwiedmann commented on bpf/lib/fib.h:
```diff
@@ -144,7 +145,7 @@ fib_do_redirect(struct __ctx_buff *ctx, const bool needs_l2_check,
 			return (int)redirect_neigh(*oif, NULL, 0, 0);
 		} else {
-			union macaddr smac = NATIVE_DEV_MAC_BY_IFINDEX(*oif);
+			union macaddr smac = device_mac(*oif);
```
From a visibility point of view, I think it would be preferable to explicitly drop the packet here if we couldn't look up the interface. And not send it out with an all-0 dst MAC.

So maybe this should be a `lookup_device()` instead?

haozhangami replied:
Done

julianwiedmann commented on bpf/tests/xdp_nodeport_lb4_nat_lb.c:
```diff
@@ -62,8 +62,10 @@ long mock_fib_lookup(__maybe_unused void *ctx, struct bpf_fib_lookup *params,
-	if (settings && settings->fail_fib)
+	if (settings && settings->fail_fib) {
+		params->ifindex = 1;
 		return BPF_FIB_LKUP_RET_NO_NEIGH;
```
Let's introduce this here as well, and refer to it further below:

```suggestion
		params->ifindex = DEFAULT_IFACE;
		return BPF_FIB_LKUP_RET_NO_NEIGH;
```

haozhangami replied:
Done.

Originally, `NATIVE_DEV_MAC_BY_IFINDEX` had a default mac. This is to lookup the smac of the interface.
```
#define NATIVE_DEV_MAC_BY_IFINDEX(_) { .addr = { 0xce, 0x72, 0xa7, 0x03, 0x88, 0x56 } }
```

Without DEFAULT_IFACE, the `xdp_nodeport_nat_fwd_reply_no_fib` test will fail. the error is `src MAC is not the LB MAC`.

julianwiedmann reviewed (commented):
Thank you! Took just a casual first look, with a few comments inline.

dylandreimerink commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -345,6 +352,21 @@ func (dc *devicesController) initialize() error {
+				}
+			}
+		})
+
```
We shouldn't manage the device map directly from this device controller. The whole point of this logic is to populate the StateDB tables which then can be consumed in a decoupled way.

This logic should be moved to https://github.com/cilium/cilium/blob/main/pkg/datapath/orchestrator/orchestrator.go

It is the job of the orchestrator to initialize the datapath on startup and to re-initialize when something like the set of current devices changes (it consumes the devices StateDB table for this purpose).

It should consume the table and reconcile the map as part of the (re)initialization process.

haozhangami replied:
This logic has been moved to Reinitialize method in the loader module. Please take a look, Thanks.

dylandreimerink commented on pkg/datapath/loader/base.go:
```diff
@@ -462,6 +507,10 @@ func (l *loader) Reinitialize(ctx context.Context, lnc *datapath.LocalNodeConfig
+	if err := l.reinitializeDeviceMap(ctx, lnc); err != nil {
+		return err
+	}
+
```
There is a specific reason I mentioned putting it in the orchestrator. There is an ongoing effort from sig-loader to simplify the logic in `pkg/datapath/loader` down to a set of mostly stateless functions for loading and attaching programs. (still in progress)

The orchestrator exists as a layer above that to deal with all of the statefull logic, doing things like:
* Waiting for other components to provide all config needed to (re) initialize the datapath
* Enforcing ordering (first reinitialize, then load LXC program)
* Re-initializating when the local node config changes (devices added/removed, runtime config changes, ect.)

So I would like to move the map sync logic into here
https://github.com/cilium/cilium/blob/main/pkg/datapath/orchestrator/orchestrator.go#L267

At that location you have the same info available as would be passed to the loader, so it should just be a case of moving it.

haozhangami replied:
This logic has been moved to the orchestrator. Thanks for the explanation!

dylandreimerink commented on pkg/maps/devicemap/devicemap.go:
```diff
@@ -0,0 +1,128 @@
+}
+
+type deviceMap struct {
+	bpfMap *ebpf.Map
```
We want to actually get rid of the `pkg/ebpf` package. Could you refactor to use the `pkg/bpf` package instead, which should have all features needed. https://github.com/cilium/cilium/issues/33128

haozhangami replied:
Done

dylandreimerink reviewed (changes requested):
I am so sorry for the delay in reviewing. The PR is generally in a good state, just one thing remaining.

Thank you so much for the work that has been put into this. It is very much appreciated.

haozhangami committed:
bpf,map: Use device map instead of device macro

This patch reimplements IS_L3_DEV and NATIVE_DEV_MAC_BY_IFINDEX using
bpf map, including the following changes:

1.Define a cilium_device_map, this new map will map device ifindexes to
  their MACs and check if they are L3 devices.
2.Orchestrator manages device entries in the above map.
3.Use is_l3_device() and device_mac() methods to replace IS_L3_DEV and
NATIVE_DEV_MAC_BY_IFINDEX

Fixes: #39116

Signed-off-by: Hao Zhang <hao.zhang.am.i@gmail.com>
